### ISSUE TITLE
PaddleNLP/examples/model_compression/ofa/run_glue_ofa.py / 代码疑问以及训练结果没有达到压缩模型效果

### ISSUE THREAD

**Issue by AI-Mart** [PaddlePaddle/PaddleNLP#1250]
for width_mult in args.width_mult_list:
    # Step8: Broadcast supernet config from width_mult,
    # and use this config in supernet training.
    net_config = utils.dynabert_config(ofa_model, width_mult)
    ofa_model.set_net_config(net_config)
    logits, teacher_logits = ofa_model(
        input_ids, segment_ids, attention_mask=[None, None])
    rep_loss = ofa_model.calc_distill_loss()
    if args.task_name == 'sts-b':
        logit_loss = paddle.zeros(shape=[1], dtype='float32')
    else:
        logit_loss = soft_cross_entropy(logits,
                                        teacher_logits.detach())
    loss = rep_loss + args.lambda_logit * logit_loss
    loss.backward()
optimizer.step()
lr_scheduler.step()
optimizer.clear_grad()

这上面是每种子网络都计算损失，然后这个损失把梯度计算出来后没有更新梯度，到for循环结束，梯度更新，这个更新的梯度是以所有子网络累计的梯度来更新吗？这么做的原因是什么？

for idx, width_mult in enumerate(args.width_mult_list):
    net_config = utils.dynabert_config(ofa_model, width_mult)
    ofa_model.set_net_config(net_config)
    tic_eval = time.time()
    if args.task_name == "mnli":
        acc = evaluate(ofa_model, criterion, metric,
                       dev_data_loader_matched, width_mult)
        evaluate(ofa_model, criterion, metric,
                 dev_data_loader_mismatched, width_mult)
        print("eval done total : %s s" %
              (time.time() - tic_eval))
    else:
        eval_result, eval_loss = evaluate(ofa_model, criterion, metric, dev_data_loader, width_mult)
        print("eval done total : %s s" %
              (time.time() - tic_eval))

    if paddle.distributed.get_rank() == 0:
        best_eval_result = eval_result
        # output_dir = args.output_dir
        output_dir = os.path.join(args.output_dir, "model_%d" % global_step)
        if not os.path.exists(output_dir):
            os.makedirs(output_dir)
        # need better way to get inner model of DataParallel
        model_to_save = model._layers if isinstance(
            model, paddle.DataParallel) else model
        model_to_save.save_pretrained(output_dir)
        tokenizer.save_pretrained(output_dir)

每个子网络都会去评估损失，然后保存模型，那保存的模型是最优子网络的模型参数吗？这里面怎么判断的呢，感觉好像是保存了最后一个子网络的模型参数？

源代码原封不动，训练完成后，相比之前的教师模型并没有达到压缩效果都是400多M这是什么原因？

**Comment by ceci3** [comment]
保存的模型是超网络的模型参数，每达到一定的step都会保存一次，所以训练过程中保存的模型参数不会比teacher少，调用export_model.py导出子模型的参数才是最终的子模型的模型参数。

**Comment by AI-Mart** [comment]
好的，感谢，也就是说导出的时候，需要参考模型压缩训练时候的日志，选择日志里面效果最佳的那个子模型，然后导出来，是这样吧

**Comment by ceci3** [comment]
是这样的



发自我的iPhone


------------------ 原始邮件 ------------------
发件人: AI-Mart ***@***.***&gt;
发送时间: 2021年11月4日 10:32
收件人: PaddlePaddle/PaddleNLP ***@***.***&gt;
抄送: ceci3 ***@***.***&gt;, Assign ***@***.***&gt;
主题: 回复：[PaddlePaddle/PaddleNLP] PaddleNLP/examples/model_compression/ofa/run_glue_ofa.py / 代码疑问以及训练结果没有达到压缩模型效果 (Issue #1250)





 
好的，感谢，也就是说导出的时候，需要参考模型压缩训练时候的日志，选择日志里面效果最佳的那个子模型，然后导出来，是这样吧
 
—
You are receiving this because you were assigned.
Reply to this email directly, view it on GitHub, or unsubscribe.
Triage notifications on the go with GitHub Mobile for iOS or Android.

**Comment by AI-Mart** [comment]
这上面是每种子网络都计算损失，然后这个损失把梯度计算出来后没有更新梯度，到for循环结束，梯度更新，这个更新的梯度是以所有子网络累计的梯度来更新吗？这么做的原因是什么？

**Comment by syc22** [comment]
> 保存的模型是超网络的模型参数，每达到一定的步骤都保存一次，所以训练过程中保存的模型参数不会比老师少，调用export_model.py导出子模型的参数才是最终的子模型的模型参数。

老师您好，export导出的模型 分为两种一种是动态.pdparams  一种是静态模型，请问导出的动态模型为什么再按照加载模型的方法XXXForSequenceClassification.from_pretrained("export_path")，结果出了预测问题，已经严重的不如裁剪前的原始模型，但是在裁剪过程中看log精度还是很高的，请问是什么原因？ 虽然官方api说明通过export导出的模型可以进一步训练，但实际上在上一步ofa_glue.py导出的超网络就已经严重不如剪裁前的原始模型了，export导出子模型后依旧严重不如。
![aaa片_20211217123956](https://user-images.githubusercontent.com/57344752/146489961-17d4ef84-a36a-494a-90ae-de9f7f65cf1d.png)
